File: frontend/js/views/paycheck_401k.js
// 401(k) Per-Paycheck Maximizer — the even deferral to max by year-end,
// and the match you forfeit by front-loading, via /calc/paycheck-401k.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';

const FIELDS = [
    ['annual_limit_usd', 'Annual deferral limit ($)', 23000],
    ['ytd_contributed_usd', 'Contributed YTD ($)', 0],
    ['pay_periods_remaining', 'Pay periods remaining', 24],
    ['gross_per_period_usd', 'Gross pay per period ($)', 4000],
    ['employer_match_pct', 'Employer match (¢ per $)', 50],
    ['match_limit_pct_of_pay', 'Match up to (% of pay)', 5],
    ['planned_per_period_usd', 'Planned deferral per period ($, 0 = skip)', 0],
];
const INT_FIELDS = new Set(['pay_periods_remaining']);

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderPaycheck401k(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.paycheck401k.h1.title">// 401(k) PER-PAYCHECK MAXIMIZER</span></h1>
        <p class="muted small" data-i18n="view.paycheck401k.hint.intro">
            How much per paycheck to hit the annual 401(k) limit by December — and whether
            you're about to leave free money on the table. Most plans match per paycheck, so
            front-loading hits the limit early and stops your contributions (and the match)
            for the rest of the year, unless the plan trues up the missed match. Spreading
            evenly captures the full match every period.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.paycheck401k.h2.inputs">Your numbers</h2>
            <form id="pc-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.paycheck401k.label.${key}">${label}</span>
                        <input type="number" step="${INT_FIELDS.has(key) ? '1' : '0.01'}" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                <label data-tip="view.paycheck401k.tip.true_up"><input type="checkbox" name="plan_has_true_up"> <span data-i18n="view.paycheck401k.label.true_up">Plan trues up match</span></label>
                <button class="primary" type="submit" data-i18n="view.paycheck401k.btn.run">Calculate</button>
            </form>
        </div>
        <div id="pc-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#pc-form');
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        const body = { plan_has_true_up: fd.get('plan_has_true_up') != null };
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        try {
            const r = await api.calcPaycheck401k(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.paycheck401k.toast.error'), { level: 'error' });
        }
    });
    form.dispatchEvent(new Event('submit'));
}

function renderResult(mount, r) {
    const el = mount.querySelector('#pc-result');
    const forfeited = Number(r.forfeited_match_usd);
    const ptl = r.periods_to_limit_at_planned;
    let note;
    if (forfeited > 0) {
        note = `<span class="neg">${t('view.paycheck401k.note.forfeit', { amount: money(forfeited), periods: r.empty_periods })}</span>`;
    } else if (ptl != null && r.empty_periods === 0) {
        note = `<span class="pos" data-i18n="view.paycheck401k.note.ok">Your planned deferral spreads to year-end — full match every check.</span>`;
    } else if (ptl != null) {
        note = `<span class="pos" data-i18n="view.paycheck401k.note.trueup">You hit the limit early, but the plan's true-up pays the missed match.</span>`;
    } else {
        note = `<span data-i18n="view.paycheck401k.note.even">Defer the even amount to capture the match every period.</span>`;
    }
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.paycheck401k.h2.result">The plan</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.paycheck401k.card.remaining">Remaining to limit</div>
                    <div class="value">${money(r.remaining_to_limit_usd)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.paycheck401k.card.even">Defer per period</div>
                    <div class="value">${money(r.even_per_period_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.paycheck401k.card.pct">% of each check</div>
                    <div class="value">${Number(r.even_pct_of_pay).toFixed(1)}%</div></div>
                <div class="card"><div class="label" data-i18n="view.paycheck401k.card.threshold">Threshold for full match</div>
                    <div class="value">${money(r.match_threshold_per_period_usd)}/check</div></div>
                <div class="card pos"><div class="label" data-i18n="view.paycheck401k.card.match">Match per check</div>
                    <div class="value">${money(r.full_match_per_period_usd)}</div></div>
                <div class="card ${forfeited > 0 ? 'neg' : ''}"><div class="label" data-i18n="view.paycheck401k.card.forfeit">Match forfeited (front-load)</div>
                    <div class="value ${forfeited > 0 ? 'neg' : ''}">${money(forfeited)}</div></div>
            </div>
            <p class="muted small">${note}</p>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/paycheck_401k.rs
//! 401(k) per-paycheck maximizer — and the front-loading match trap.
//!
//! Two questions every year: how much per paycheck do I defer to hit the
//! annual limit by December, and am I about to forfeit free money? The
//! trap: most plans match per paycheck, so if you front-load and hit the
//! limit early, you contribute $0 on the remaining checks and the
//! employer matches $0 on them too — unless the plan has a year-end
//! "true-up" that pays the missed match. Spreading evenly captures the
//! full match every period.
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Paycheck401kInput {
    /// Employee elective-deferral limit for the year.
    pub annual_limit_usd: f64,
    /// Already deferred year-to-date.
    pub ytd_contributed_usd: f64,
    pub pay_periods_remaining: u32,
    pub gross_per_period_usd: f64,
    /// Employer match rate, cents per dollar (e.g. 50 = $0.50 per $1).
    pub employer_match_pct: f64,
    /// Employer matches deferrals up to this percent of pay per period.
    pub match_limit_pct_of_pay: f64,
    /// Does the plan true-up missed match at year-end?
    pub plan_has_true_up: bool,
    /// What you're currently set to defer per period — for the
    /// front-loading check (0 to skip it).
    pub planned_per_period_usd: f64,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct Paycheck401kResult {
    pub remaining_to_limit_usd: f64,
    /// Defer this per period to hit the limit exactly at year-end.
    pub even_per_period_usd: f64,
    pub even_pct_of_pay: f64,
    /// Defer at least this per period (pay × match-limit%) for full match.
    pub match_threshold_per_period_usd: f64,
    /// Employer dollars per period at or above the threshold.
    pub full_match_per_period_usd: f64,
    /// Does the even-spread deferral also clear the match threshold?
    pub captures_full_match_evenly: bool,
    /// Periods until the planned deferral hits the limit (`None` if no
    /// planned amount given).
    pub periods_to_limit_at_planned: Option<u32>,
    /// Periods you'd contribute $0 after hitting the limit early.
    pub empty_periods: u32,
    /// Match forfeited by front-loading without a true-up (else 0).
    pub forfeited_match_usd: f64,
}

pub fn compute(i: &Paycheck401kInput) -> Paycheck401kResult {
    let remaining = (i.annual_limit_usd - i.ytd_contributed_usd).max(0.0);
    let periods = i.pay_periods_remaining;

    let even_per_period = if periods > 0 { remaining / periods as f64 } else { 0.0 };
    let even_pct_of_pay = if i.gross_per_period_usd > 0.0 {
        even_per_period / i.gross_per_period_usd * 100.0
    } else {
        0.0
    };

    let threshold = i.gross_per_period_usd * (i.match_limit_pct_of_pay / 100.0);
    let full_match_per_period = (i.employer_match_pct / 100.0) * threshold;
    let captures_full_match_evenly = even_per_period >= threshold;

    let (periods_to_limit, empty_periods) = if i.planned_per_period_usd > 0.0 && remaining > 0.0 {
        let k = (remaining / i.planned_per_period_usd).ceil() as u32;
        let k = k.min(periods);
        (Some(k), periods.saturating_sub(k))
    } else {
        (None, 0)
    };

    let forfeited_match = if i.plan_has_true_up {
        0.0
    } else {
        full_match_per_period * empty_periods as f64
    };

    Paycheck401kResult {
        remaining_to_limit_usd: remaining,
        even_per_period_usd: even_per_period,
        even_pct_of_pay,
        match_threshold_per_period_usd: threshold,
        full_match_per_period_usd: full_match_per_period,
        captures_full_match_evenly,
        periods_to_limit_at_planned: periods_to_limit,
        empty_periods,
        forfeited_match_usd: forfeited_match,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> Paycheck401kInput {
        Paycheck401kInput {
            annual_limit_usd: 23_000.0,
            ytd_contributed_usd: 0.0,
            pay_periods_remaining: 24,
            gross_per_period_usd: 4_000.0,
            employer_match_pct: 50.0,      // $0.50 per $1
            match_limit_pct_of_pay: 5.0,   // up to 5% of pay
            plan_has_true_up: false,
            planned_per_period_usd: 0.0,
        }
    }

    #[test]
    fn even_spread_hits_limit() {
        let r = compute(&base());
        // $23,000 / 24 ≈ $958.33 per period.
        assert!((r.even_per_period_usd - 23_000.0 / 24.0).abs() < 1e-6);
        assert!((r.even_pct_of_pay - (23_000.0 / 24.0) / 4_000.0 * 100.0).abs() < 1e-6);
    }

    #[test]
    fn match_threshold_and_full_match() {
        let r = compute(&base());
        // 5% of $4,000 = $200 to earn full match; 50% of $200 = $100.
        assert!((r.match_threshold_per_period_usd - 200.0).abs() < 1e-9);
        assert!((r.full_match_per_period_usd - 100.0).abs() < 1e-9);
        // Even spread ($958) clears the $200 threshold → full match all year.
        assert!(r.captures_full_match_evenly);
    }

    #[test]
    fn front_loading_forfeits_match_without_true_up() {
        // Defer $2,000/period → hits $23k after 12 periods, idle for 12,
        // forfeiting $100 match × 12 = $1,200.
        let r = compute(&Paycheck401kInput { planned_per_period_usd: 2_000.0, ..base() });
        assert_eq!(r.periods_to_limit_at_planned, Some(12));
        assert_eq!(r.empty_periods, 12);
        assert!((r.forfeited_match_usd - 1_200.0).abs() < 1e-9);
    }

    #[test]
    fn true_up_plan_forfeits_nothing() {
        let r = compute(&Paycheck401kInput {
            planned_per_period_usd: 2_000.0,
            plan_has_true_up: true,
            ..base()
        });
        assert_eq!(r.empty_periods, 12); // still hits early...
        assert_eq!(r.forfeited_match_usd, 0.0); // ...but the true-up pays it back
    }

    #[test]
    fn planned_within_limit_no_front_load() {
        // $958/period over 24 periods exactly hits the limit — no idle periods.
        let r = compute(&Paycheck401kInput { planned_per_period_usd: 958.34, ..base() });
        assert_eq!(r.empty_periods, 0);
        assert!((r.forfeited_match_usd - 0.0).abs() < 1e-9);
    }

    #[test]
    fn already_maxed_leaves_nothing() {
        let r = compute(&Paycheck401kInput { ytd_contributed_usd: 23_000.0, ..base() });
        assert_eq!(r.remaining_to_limit_usd, 0.0);
        assert_eq!(r.even_per_period_usd, 0.0);
    }

    #[test]
    fn no_planned_skips_front_load_check() {
        let r = compute(&base());
        assert_eq!(r.periods_to_limit_at_planned, None);
        assert_eq!(r.empty_periods, 0);
        assert_eq!(r.forfeited_match_usd, 0.0);
    }
}
